axum-extra/src/ndjson.rs

```diff
@@ -0,0 +1,288 @@
+//! Newline delimited JSON extractor and response.
+
+use axum::{
+    async_trait,
+    body::{HttpBody, StreamBody},
+    extract::{rejection::BodyAlreadyExtracted, FromRequest, RequestParts},
+    response::{IntoResponse, Response},
+    BoxError,
+};
+use bytes::{BufMut, Bytes, BytesMut};
+use futures_util::stream::{BoxStream, Stream, TryStream, TryStreamExt};
+use http::header::CONTENT_TYPE;
+use pin_project_lite::pin_project;
+use serde::{de::DeserializeOwned, Serialize};
+use std::{
+    io::{self, Write},
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tokio::io::AsyncBufReadExt;
+use tokio_stream::wrappers::LinesStream;
+use tokio_util::io::StreamReader;
+
+pin_project! {
+    /// A stream of newline delimited JSON ([NDJSON]).
+    ///
+    /// This can be used both as an extractor and as a response.
+    ///
+    /// # As extractor
+    ///
+    /// ```rust
+    /// use axum_extra::ndjson::NdJson;
+    /// use futures::stream::StreamExt;
+    ///
+    /// async fn handler(mut stream: NdJson<serde_json::Value>) {
+    ///     while let Some(value) = stream.next().await {
+    ///         // ...
+    ///     }
+    /// }
+    /// ```
+    ///
+    /// # As response
+    ///
+    /// ```rust
+    /// use axum::{BoxError, response::{IntoResponse, Response}};
+    /// use axum_extra::ndjson::NdJson;
+    /// use futures::stream::Stream;
+    ///
+    /// fn stream_of_values() -> impl Stream<Item = Result<serde_json::Value, BoxError>> {
+    ///     # futures::stream::empty()
+    /// }
+    ///
+    /// async fn handler() -> Response {
+    ///     NdJson::new(stream_of_values()).into_response()
+    /// }
+    /// ```
+    ///
+    /// [NDJSON]: https://ndjson.org/
+    // we use `AsExtractor` as the default because you're more likely to name this type if its used
+    // as an extractor
+    pub struct NdJson<S, T = AsExtractor> {
+        #[pin]
+        inner: Inner<S>,
+        _marker: PhantomData<T>,
+    }
+}
+
+pin_project! {
+    #[project = InnerProj]
+    enum Inner<S> {
+        Response {
+            #[pin]
+            stream: S,
+        },
+        Extrator {
```

Comment:
```suggestion
        Extractor {
```

Reply:
I'll fix the compile errors this will give later :)